### ISSUE TITLE
Switch to Arrow as data container

### ISSUE THREAD

**Issue by jkiviluo** [Spine-project/Spine-Toolbox#2506]
Currently parameter data in Spine DBs is held in JSON based binary blobs. While it has given us the flexibility we needed, it's not the most performant option (especially nowadays with more options available). Arrow looks like a very nice fit for our purposes. It supports all the datatypes (https://arrow.apache.org/docs/cpp/api/datatype.html) we currently have, but will handle dataframe type data much faster and more memory efficiently. Pandas has started to support Arrow (instead of NumPy) (https://datapythonista.me/blog/pandas-20-and-the-arrow-revolution-part-i). There is also native SQL support (Flight SQL), but how we could make that work needs to be explored (but could give a very large speed boost when accessing data from DBs).

Inspired by https://github.com/spine-tools/Spine-Toolbox/issues/2497#issuecomment-1895063202 and discussion with @soininen

**Comment by jkiviluo** [comment]
COMPARISON TABLE EDITED 13.8. 17:40 (added steel-man JSON)

I started looking into this, since I would really like to see some speedup in the DB editor (and DB generally).

The current WIP is aiming at tabular JSON, but Claude benchmarked that:

1,000 × 8,760-pt series, real SQLite bulk read           │ wall time
Old JSON blob                            │ 10.9 s
Branch tabular JSON (pydantic) │ 21.3 s
Steel-man fast-JSON reader | 2.38 s
Arrow IPC stream                      │ 87.5 ms
Arrow IPC + zstd                        │ 285 ms  (AND 3-3.5x smaller database)
  
The recommendation was a hybrid: "keep the branch's entire logical model, facade, migration harness, and tests (all encoding-agnostic), but swap the physical encoding of indexed types (array, time_series, time_pattern, map, table) to Arrow IPC stream bytes in the existing LargeBinary column, while scalars/date_time/duration stay tiny JSON (IPC's ~288-byte framing is strictly worse for scalars).  The existing type column plus IPC magic-byte sniffing handles format dispatch and backward compatibility. Bonus: Julia's Arrow.jl reads IPC natively, so SpineInterface's parser task actually shrinks compared to parsing the bespoke tabular JSON."

What do you think? I can guide Claude to implement IPC + zstd (zstandard compression algorithm, comes out of IPC box), if you agree with that.

**Comment by jkiviluo** [comment]
Here is a high-level implementation plan generated by Claude (with decision points, risks and validation strategy)

[arrow-transition-plan-v2.md](https://github.com/user-attachments/files/31034290/arrow-transition-plan-v2.md)

**Comment by soininen** [comment]
@suvayu you may want to chime in here as well.

Which Arrow IPC are we talking about? [The dissociated IPC](https://arrow.apache.org/docs/format/DissociatedIPC.html)? That protocol is marked as experimental, so we may want to wait until it is fully standardized. The binary blobs must have a stable format - we do not want to provide new migrations every time they update the protocol which is expected to happen often while it is still in experimental phase.

That being said, I'm not against us revisiting the binary format we store in the database, or trying out other JSON parsers.

As a little side note, that implementation plan is not high-level nor easy to read. Next time, please provide an easy to comprehend, human-written plan.

**Comment by jkiviluo** [comment]
Fair enough, here is the plan (human-edited) condensed to plain English:

**Goal:** Store indexed parameter values (time series, maps, arrays, time patterns) as compact binary instead of JSON text, so they're smaller on disk and much faster to read and write. Scalars and small values stay as they are.

**Which binary format:** The standard Arrow IPC stream format — the same one pyarrow writes and Arrow.jl reads. Not the experimental Dissociated IPC protocol, which is unrelated.

**The one caveat:** The format is stable, but the exact bytes for the same value can differ slightly between pyarrow/zstd versions. That matters when the code compares value blobs byte-for-byte. The one place where we need to be careful is list_value_lookup, which will be handled with a simple rule: *list values stay JSON*.   A footnote: the one other blob comparison (update diffing) already falls back to comparing parsed values, so it doesn't care whether bytes are stable. Nothing else in any repo compares blobs. So Arrow-version byte drift can't cause a correctness bug; canonical python-based writer and drift-CI are optional polish, not requirements.

**What stays human-readable:** Copy/paste and hand-editing of values must keep working. So the clipboard/interchange text form stays JSON (a clean tabular JSON), even though storage is binary.

**Tiny values stay JSON:** Below ~20 rows, JSON is actually smaller and just as fast, so short values are kept as text. The rule is purely on row count, so the same value always encodes the same way.

**Migration:** Existing databases convert once, in a single migration that's chunked, restartable, and re-runnable (already-converted rows are skipped). Backups are strongly urged as before. Downgrades aren't supported.

**Julia side:** SpineInterface reads these blobs directly with Arrow.jl (its hand-rolled JSON parser can be buried) and writes results as IPC too. Julia's bytes don't need to be pixel-perfect because the Python server re-encodes everything on import. The client/server wire protocol gets a small framing fix so old and new versions negotiate cleanly instead of hanging.

**What we're giving up:** You lose SQL json_each-style querying into indexed values (scalars stay queryable). If that matters, we can add an arrow_to_json helper. This is the same trade-off discussed in #2682 — the difference now is we've measured it: binary is dramatically faster to read and write and meaningfully smaller, and the JSON logical schema survives as the interchange format, so nothing about the data model is thrown away.

**Comment by soininen** [comment]
Just to make sure we discuss the same format, we are talking about [this](https://arrow.apache.org/docs/format/Columnar.html#ipc-streaming-format), right?

**Comment by jkiviluo** [comment]
Yes. And arrow IPC + zstd is still the same format (the compression is in-spec feature of the IPC message body).

**Comment by jkiviluo** [comment]
## Benchmark: JSON vs SQLite jsonb vs Arrow IPC for parameter values
EDITED 1h after posting: added fixed-resolution comparison too.


After our telco discussion, I had Claude measure the encodings across the access patterns.

Time series come in two flavours that behave very differently, so I've split them.
**Fixed-resolution** (start + resolution + a values array, *no stored timestamps*) is
the compact, common case — and the one with a migration trap. **Variable-resolution**
(explicit timestamps) is where Arrow dominates. Fixed first.

**Setup:** SQLite 3.45.1, pyarrow 24.0.0, numpy 1.26.4; 8760-pt series (500 series for
the bulk/cross-table numbers); current formats per the parameter_value spec; warm cache;
best-of runs. All object-equality / cross-table / first-100 results were cross-checked
**identical** across encodings before timing.

### 1. Fixed-resolution time series

No timestamps are stored — just `start`, `resolution` and a values array — so JSON is
already compact (~4× smaller than variable-resolution) and there is no per-timestamp
parse on read. Arrow's advantage here is therefore **modest, and conditional**.

Storage, one 8760-pt series:

| encoding | size |
|---|--:|
| JSON (values only) | 67.6 KB |
| Arrow values + metadata | 68.9 KB |
| Arrow values + metadata + zstd | **31.3 KB** |
| Arrow *collapsed* to (timestamp, value) | ⚠️ 137 KB |

Deserialise to a `ParameterValue` (`TimeSeriesFixedResolution`); the index is lazy:

| | JSON | Arrow |
|---|--:|--:|
| construct (index not materialised) | 724 µs | **24 µs** |
| construct + access `.indexes` (plot / edit) | 1.31 ms | **599 µs** (~2.2×) |

- **Storage is a wash without compression** — raw Arrow ≈ JSON (both dominated by the
  same float values). zstd is what buys the ~2× (31 vs 68 KB).
- **The realistic editor ratio is ~2×, not an order of magnitude.** Once you plot or
  edit (access `.indexes`), both formats pay the same ~575 µs to generate timestamps and
  build the index lookup dict — a shared floor. The 40× construct-only ratio only holds
  while the index is never touched.
- **⚠️ Do not "collapse" fixed → variable.** Materialising `(timestamp, value)` columns
  is **2× larger than today's JSON** (137 vs 68 KB) and throws away the
  `TimeSeriesFixedResolution` identity. This is decision **D1**: the codec must store
  fixed-res compactly (values + `start`/`resolution` in schema metadata). The current
  branch tests deliberately collapse — that would make fixed-resolution data *bigger*,
  not smaller.

### 2. Variable-resolution — every access pattern

Explicit timestamps: JSON stores 8760 ISO strings and parses each one on read. This is
where Arrow wins across the board (500 series × 8760 pts; fastest per row in **bold**;
reads decode to numpy arrays — the real editor cost is section 4).

| Access pattern | json_text | jsonb | arrow_ipc | ipc+zstd |
|---|--:|--:|--:|--:|
| Storage (per series) | 278 KB | 252 KB | 140 KB | **45.5 KB** |
| Write + store (all) | 1.54 s | 1.86 s | **60 ms** | 173 ms |
| Whole-value read (all → arrays) | 1.06 s | 1.16 s | **34.5 ms** | 79.7 ms |
| First 100 pts (one series) | 329 µs | 85 µs | **51 µs** | 147 µs |
| Value @ one time, all series | 324 ms | 201 ms | **43 ms** | 88 ms |
| Average, all points/series | 526 ms | 395 ms | **41 ms** | 82 ms |
| Length of every series (tooltip) | 224 ms | 97 ms | **30 ms** | 77 ms |

**The surprise — the two SQL-query rows also lose.** `json_each` over a value-keyed
blob has no index into the content, so it *fully parses every blob* to filter or
aggregate. JSON parsing is ~30× slower than Arrow's columnar decode, so "pull all
blobs + decode + aggregate in numpy" beats "aggregate in SQL" by 7–13×. In-DB query
doesn't avoid materialisation — it relocates it into a slower parser. Genuine in-DB
querying would need a normalised, indexed `(series, time, value)` table — i.e.
abandoning blobs, not "queryable JSON".

### 3. "First 100 rows" — does anything do a true partial read?

(Independent of resolution — this is about slicing into one value.) Slice fixed at 100
rows; series length grown 1k → 1M. Flat cost = a real partial read; rising cost = the
whole blob is being touched.

| Series length | json_text | jsonb | ipc 1-batch | ipc chunked | ipc + blob I/O |
|---|--:|--:|--:|--:|--:|
| 1,000 | 49.8 µs | 23.6 µs | 15.6 µs | 15.2 µs | **13.2 µs** |
| 10,000 | 297 µs | 30.5 µs | 23.3 µs | 23.3 µs | **13.5 µs** |
| 100,000 | 2.94 ms | 97.6 µs | 86.9 µs | 94.9 µs | **13.3 µs** |
| 1,000,000 | 32.8 ms | 2.81 ms | 3.33 ms | 3.76 ms | **14.0 µs** |

`json_each … LIMIT 100` is a full parse — the worst option, and it gets worse as the
series grows. jsonb and single-/chunked-IPC all converge to ~3 ms at 1M because the
**whole cell is fetched** regardless. The only true partial read is **chunked Arrow +
SQLite incremental blob I/O**, staying flat (~2,300× faster at 1M). So "partial read
of the first N rows" is an Arrow capability, not a JSON one.

**Proof it's real on SQLite** — one 17.9 MB value, `blobopen` reads only a 16 KB prefix:

```
full blob bytes : 17,920,176
bytes read      :     16,384   (0.09% of blob)
first batch     : 100 rows, correct values
partial (blobopen prefix) :    12.9 µs
whole cell (SELECT)       : 19,994 µs   (~1,550×, identical values)
```

`sqlite3.Connection.blobopen()` (Python 3.11+) over the `sqlite3_blob_open/read` C
API. SQLite-only, needs the raw connection + rowid and a chunked value; Postgres/MySQL
fall back to whole-cell fetch. Not something the current single-batch stream plan
gets for free — it's a deliberate add-on.

### 4. Variable-resolution editor cost: deserialising to a value object

The DB editor turns every blob into a `TimeSeriesVariableResolution`, not a numpy
array. Real `from_database()` vs an Arrow codec producing the identical object
(8760 pts, verified equal):

| | JSON | Arrow |
|---|--:|--:|
| **End-to-end: blob → ParameterValue** | **4.93 ms** | **526 µs (9.4×)** |
| `json.loads` | 1.22 ms | — |
| dict → ParameterValue (`from_dict`) | 3.66 ms | — |
| — of which per-element `np.datetime64` loop | 2.23 ms | — |
| IPC decode + columns → numpy | — | 8.5 µs |
| ParameterValue constructor | (incl. above) | 496 µs |

Most of the JSON cost — `json.loads` plus a per-element `np.datetime64` loop over all
8760 stamps — is exactly what Arrow removes (it feeds native arrays straight in). The
one *format-neutral* shared cost is the eager `position_lookup` dict built in
`_Indexes.__new__` (~510 µs; both paths pay it, it's the floor under Arrow — same floor
as fixed-res section 1). Making it lazy would push Arrow toward its ~10 µs decode floor.
So the app's data format compresses the win from ~30× (raw decode) to ~9.4× end-to-end —
still decisive for variable-resolution, and it relocates the remaining bottleneck to a
fixable spot in spinedb_api.

### jsonb reality check

- Only ~10% smaller than text JSON (numbers stay as text).
- Whole-value read is **slower** than text — must convert back to text for Python.
- Queries ~1.5–2× faster than text `json_each`, but still 3–5× behind Arrow.

### Caveats

Warm cache; single RecordBatch per value; no secondary indexes; SQLite only. The flat
partial-read path needs chunked batches (not in the current plan) + `blobopen`
(Postgres would need large objects). Time series shown; Map is more nested, so Arrow's
edge there should be larger. Harnesses (`bench_blob.py`, `bench_first_n.py`,
`bench_editor.py`, `bench_fixed.py`) available on request.
